docs/wiki/testing/formal-verification.md

```diff
@@ -0,0 +1,172 @@
+
+Similarly, the number `5` is an abstraction for a team; and the answer `16.6` abstractly represents the total crates of beer required, which we rounded up to `17` because we wanted to order whole crates. That we quickly perform these abstractions and interpret the result effortlessly in our minds doesn't take away from the fact that abstraction is a powerful tool.
+
+Our algorithm can be outlined as:
+
+1. For a given problem, abstract away the irrelevant details from the input.
+2. Use a mathematical model to find a solution.
+3. Map the solution back to the original context.
+
+This is a fundamental theme in problem solving. Computers do this too. The words on your screen are abstractly [represented in binary on a computer](https://en.wikipedia.org/wiki/ASCII) for transmission and storage.
+
+> Mathematics is the language of precision. Abstraction allows us to precisely define the input focused only on relevant features for a given problem. The result of which can be mathematically solved and interpreted as required.
+
+## Formal verification
+
+Formal verification, like the previous example, helps find the answer to a simple question: "Does a system correctly meet its required specifications?"
+
+> Specifically, formal verification is a verification technique that abstracts a system as a mathematical model and proves or disproves its correctness.
+
+A "system" is defined as a mechanism that is able to execute all of the functions given by its external interface. 
+
+"Invariants" are properties that remains unchanged for a system, regardless of its current state. For example, an invariant of a vending machine is: Nobody should be able to dispense a product for free.
+
+Formal verification tests the correctness of a system by checking if all its invariants holds true.
+
+The philosophical roots of formal methods reach back to ancient Greece with Plato's exploration of theory of forms in his book "Sophist", while mathematicians in the 17th century further developed the concept through abstract algebra. German polymath, Gottfried Leibniz's vision laid the groundwork for what we now call formal reasoning. In the 19th century, pioneering work by George Boole on analysis and Friedrich Frege on propositional logic provided the foundation for formal semantics.
+
+There are several types of formal methods used to verify a system design:
+
+- **Model checking / assertion-based checking**: Models a system as a finite state machine and verifies its correctness and liveness using propositional logic.
+- **Temporal logic**: Models a system whose propositions varies with time.
+- **Equivalence checking**: Verifies if two models of the same specification but varying implementations produce the same result.
+
+Popular tools verification systems are:
```

Comment:
Maybe add more context about these tools, what are they used for, etc because it's not ethereum specific

Reply:
Also one important thing missing here is coq, older and maybe the most adopted tool